### ISSUE TITLE
Feature request: Ed448-Goldilocks

### ISSUE THREAD

**Issue by tarcieri** [jedisct1/libsodium#254]
[Ed448-Goldilocks](http://ed448goldilocks.sourceforge.net/) was recently selected by the IRTF CFRG to be the next-generation high security elliptic curve. While Curve25519/Ed25519 are probably secure for most purposes, some people like to tout extra security bells and whistles, and Ed448-Goldilocks actually delivers pretty good performance for what is probably an unnecessary margin of security ("[Spinal Tap-grade security](https://silentcircle.wordpress.com/tag/spinal-tap/)")

It'd be nice if libsodium provided an implementation of it when there is a sufficiently mature one available.

It should be usable for both ECDH (i.e. scalar multiplication) and signatures, although the CFRG has not yet standardized anything regarding signatures. My understanding from various ECC implementers is that adapting it to EdDSA should not be that difficult.

**Comment by sarciszewski** [comment]
Wait, I'm sorry if this is off-topic, but the CFRG finally stopped bickering over endianness and managed to put forth a formal recommendation? :O

**Comment by tarcieri** [comment]
Yes, they [announced it back in March](https://mailarchive.ietf.org/arch/msg/cfrg/BPDuOnVbrWiMSCjcfP6YTiaS6NY)

**Comment by dconnolly** [comment]
Adding my :+1: to this. Does this need to go into NaCl first?

**Comment by cfcs** [comment]
A :+1: from me too :-)

**Comment by fefafefa** [comment]
A :+1: too!  :D

**Comment by mitzip** [comment]
:+1:

**Comment by gpalomar** [comment]
![image](https://cloud.githubusercontent.com/assets/3463553/11456671/7335afc6-9690-11e5-9b1a-fc31dcc81a70.png)

**Comment by gpalomar** [comment]
:+1:

**Comment by tarcieri** [comment]
@bitwiseshiftleft is there an Ed448 implementation sufficiently stable for inclusion in libsodium yet?

**Comment by paragonie-scott** [comment]
:-1: https://twitter.com/jedisct1/status/691555098766544900 :P

**Comment by potatosalad** [comment]
For those looking for a reference implementation of Ed448, Mike Hamburg has been working on [libdecaf](https://sourceforge.net/projects/ed448goldilocks/) ([mirror here](https://github.com/potatosalad/ed448goldilocks)) which has mostly stabilized since [RFC 7748](https://tools.ietf.org/html/rfc7748) has been finalized and the [EdDSA draft](https://tools.ietf.org/html/draft-irtf-cfrg-eddsa-05) is almost finalized.  The experimental [erlang-libdecaf](https://github.com/potatosalad/erlang-libdecaf) shows some of the functions provided by the implementation (for example under [`c_src/ed448_funcs.h`](https://github.com/potatosalad/erlang-libdecaf/blob/master/c_src/ed448_funcs.h)).

**Comment by kangjoni76** [comment]
+1

**Comment by jamesruan** [comment]
Good to see.

**Comment by pdefreitas** [comment]
👍

**Comment by maqp** [comment]
Is the curve suitable for both Diffie-Hellman and digital signatures? If yes, is the goal to provide both?

**Comment by potatosalad** [comment]
@maqp I don't know about the goals for libsodium, but yes, Curve448 has the X448 function as defined in [RFC 7748](https://tools.ietf.org/html/rfc7748) for usage with Diffie-Hellman and Ed448 (and Ed448ph) is defined in [RFC 8032](https://tools.ietf.org/html/rfc8032) for digital signatures.

**Comment by maqp** [comment]
@potatosalad Great. That's definitely something I'd like to see implemented; Expansion of X25519's keys for use in XSalsa20 when they have only 128 bits of symmetric security feels odd. 

On the other hand, Shor's algorithm defeats exchange in polynomial time despite bit strength, so 256 bit symmetric security doesn't add much in this case. 128 bits is enough against digital electronic super computers for the next 100+ years.

**Comment by paragonie-scott** [comment]
X448 isn't impossible to add, but you're looking at bloating the public key API to twice its current size in order to accommodate X448:

* Take everything [here](https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_box/crypto_box.c) and add an `_x448` variant.
* Also add a SHAKE256 implementation, for Ed448.

If we are to have Curve448 implemented, I'd like to also see it be combined with XChaCha20Poly1305 instead of XSalsa20Poly1305, for crypto_box-like use-cases.

**Comment by YuGits** [comment]
Since you're interested in ed448, I wondered if any of you already have tried to sign and verify messages with ed448 using libdecaf? If so, what were your experiences with it?
If not and you're interested in testing libdecaf, here's a tutorial about how you can do that: [[ed448 tutorial]](https://github.com/openssl/openssl/pull/3503#issuecomment-303804857) If you find bugs in libdecaf, you can submit a [bugreport](https://sourceforge.net/p/ed448goldilocks/tickets/).

**Comment by potatosalad** [comment]
@YuGits Yes, and it works great.  I put together an Erlang C library that leverages libdecaf for Ed448 and Ed448ph: [erlang-libdecaf#ed448](https://github.com/potatosalad/erlang-libdecaf#ed448).  I'm using it in a few internal projects in JSON Web Signature (JWS) form as described in [RFC8037](https://tools.ietf.org/html/rfc8037) (here are [some examples](https://hexdocs.pm/jose/JOSE.JWS.html#module-examples) in Elixir).  X448 also works well (see [erlang-libdecaf#x448](https://github.com/potatosalad/erlang-libdecaf#x448)) and can be used in JSON Web Encryption (JWE) form (see [here](https://hexdocs.pm/jose/JOSE.JWE.html#module-key-derivation-examples) and scroll down to the ECDH-ES section).

For anyone interested, I also have pure implementations of Ed448, Ed448ph, and X448 in both Erlang and Ruby:

* Ed448 and Ed448ph: [Erlang](https://github.com/potatosalad/erlang-jose/blob/master/src/jose_jwa_ed448.erl) - [Ruby](https://github.com/potatosalad/ruby-jose/blob/master/lib/jose/jwa/ed448.rb)
* X448: [Erlang](https://github.com/potatosalad/erlang-jose/blob/master/src/jose_jwa_x448.erl) - [Ruby](https://github.com/potatosalad/ruby-jose/blob/master/lib/jose/jwa/x448.rb)

**Comment by YuGits** [comment]
@potatosalad Do you think the C implementation of X448/Ed448 you have is suitable for integration with OpenSSL?

See this quote of OpenSSL co-founder [Stephen Henson](https://github.com/openssl/openssl/pull/3503#issuecomment-303707291):
> The internal API between OpenSSL and the low level algorithm [of X25519 and Ed25519] is very simple: 5 functions covering X25519 and Ed25519.  Adding X448/Ed448 support should also be much easier by following a similar model but **we're lacking the low level algorithm code in a form we can use.**

**Comment by potatosalad** [comment]
@YuGits Mike Hamburg (@bitwiseshiftleft) would probably have a better answer than me as he's the real author of [ed448goldilocks or "libdecaf"](https://sourceforge.net/projects/ed448goldilocks/).  I just wrote the C bindings for Erlang as well as the pure Erlang and Ruby implementations based on the CFRG documents and what I learned from him.

One of the issues with Ed448 is it actually needs the SHA-3 functions available (or at least SHAKE256), so openssl/openssl#439 would need to be resolved at the same time.  X448, on the other hand, doesn't have any external algorithm dependencies.
